Add IPv4 wildcard pattern matching for the listening IP addresses

Burgos opened the pull request:
On some systems, names of the network interfaces can change over time (when installing/uninstalling adapters), so the private and the public interface might be swapped with each other. This adds the optional and simple IPv4 wildcard pattern matching to make sure qBitTorrent doesn't listen outside on the IP addresses that doesn't match the pattern.

Burgos commented:
TravisCI failure looks unrelated to this PR, and previous PRs are failing with the same issue.

thalieht commented on src/base/bittorrent/session.cpp:
```diff
@@ -2412,13 +2413,34 @@ void Session::networkConfigurationChange(const QNetworkConfiguration& cfg)
+    QStringList patternPieces = pattern.split(".");
+
+    if (ipPieces.size() != patternPieces.size())
+    {
```
Opening bracket in the same line with if's: `if () {`
The same for `else`, `else if ()` and `for ()`.

thalieht commented on src/base/bittorrent/session.cpp:
```diff
@@ -2462,6 +2500,19 @@ const QStringList Session::getListeningIPs()
+        // only for ipv4).
+        if (!ifaceAddrPattern.isEmpty() && !addressPatternMatches(ipString, ifaceAddrPattern))
+        {
+            logger->addMessage(tr("Blocked address %1 on pattern %2", "Blocked address 124.5.1568.1 on pattern 10.*.*.*").arg(ipString, ifaceAddrPattern), Log::INFO);
```
Typo:  124.5.1568.1

thalieht commented on src/base/bittorrent/session.cpp:
```diff
@@ -2432,8 +2454,24 @@ const QStringList Session::getListeningIPs()
+        {
+            if (!ifaceAddrPattern.isEmpty() && !addressPatternMatches(ifaceAddr, ifaceAddrPattern))
+            {
+                logger->addMessage(tr("Configured network interface address %1 doesn't match the given pattern.", "Configured network interface address 124.5.1568.1 doesn't match the given pattern.").arg(ifaceAddr), Log::CRITICAL);
```
Typo: 124.5.1568.1

Burgos replied:
Whoops, it's a silly copy/paste error (I've copied it from the log message for the invalid address).

thalieht commented on src/base/bittorrent/session.cpp:
```diff
@@ -2462,6 +2500,19 @@ const QStringList Session::getListeningIPs()
+        }
+        else
+        {
+            logger->addMessage(tr("Allowed address %1 on pattern %2", "Allowed address 10.5.1568.1 on pattern 10.*.*.*").arg(ipString, ifaceAddrPattern), Log::INFO);
```
Typo: 10.5.1568.1

thalieht commented on src/base/bittorrent/session.cpp:
```diff
@@ -2432,8 +2454,24 @@ const QStringList Session::getListeningIPs()
+        }
+        else if (!ifaceAddrPattern.isEmpty())
+        {
+            logger->addMessage(tr("No network interface/address is specified, but address pattern is set.", "No network interface/address is specified, but address pattern is set."), Log::CRITICAL);
```
The second argument of `tr()` is not needed because it's the same as the first.

Burgos commented:
Thank you for the comments, I'll address them during the day!

On 29 Dec 2017 14:51, "thalieht" <notifications@github.com> wrote:

*@thalieht* commented on this pull request.

Coding style comments. A few things you will be asked to change when the
real reviews begin.
------------------------------

In src/base/bittorrent/session.cpp
<https://github.com/qbittorrent/qBittorrent/pull/8159#discussion_r159061818>
:

> @@ -2412,13 +2413,34 @@ void Session::networkConfigurationChange(const QNetworkConfiguration& cfg)
     }
 }

+bool Session::addressPatternMatches(const QString& ipAddress, const
QString& pattern)
+{
+    QStringList ipPieces = ipAddress.split(".");
+    QStringList patternPieces = pattern.split(".");
+
+    if (ipPieces.size() != patternPieces.size())
+    {

Opening bracket in the same line with if's: if () {
The same for else, else if () and for ().
------------------------------

In src/base/bittorrent/session.cpp
<https://github.com/qbittorrent/qBittorrent/pull/8159#discussion_r159062123>
:

> @@ -2462,6 +2500,19 @@ const QStringList Session::getListeningIPs()
             || (listenIPv6 && (protocol == QAbstractSocket::IPv4Protocol)))
             continue;

+        // If an iface address pattern has been defined only allow
ips that match it to go
+        // through. Note that this will discard ipv6 addresses (the
wildcard matching is supported
+        // only for ipv4).
+        if (!ifaceAddrPattern.isEmpty() &&
!addressPatternMatches(ipString, ifaceAddrPattern))
+        {
+            logger->addMessage(tr("Blocked address %1 on pattern %2",
"Blocked address 124.5.1568.1 on pattern 10.*.*.*").arg(ipString,
ifaceAddrPattern), Log::INFO);

Typo: 124.5.1568.1
------------------------------

In src/base/bittorrent/session.cpp
<https://github.com/qbittorrent/qBittorrent/pull/8159#discussion_r159062136>
:

> @@ -2432,8 +2454,24 @@ const QStringList Session::getListeningIPs()

     if (ifaceName.isEmpty()) {
         if (!ifaceAddr.isEmpty())
-            IPs.append(ifaceAddr);
-        else
+        {
+            if (!ifaceAddrPattern.isEmpty() &&
!addressPatternMatches(ifaceAddr, ifaceAddrPattern))
+            {
+                logger->addMessage(tr("Configured network interface
address %1 doesn't match the given pattern.", "Configured network
interface address 124.5.1568.1 doesn't match the given
pattern.").arg(ifaceAddr), Log::CRITICAL);

Typo: 124.5.1568.1
------------------------------

In src/base/bittorrent/session.cpp
<https://github.com/qbittorrent/qBittorrent/pull/8159#discussion_r159062160>
:

> @@ -2462,6 +2500,19 @@ const QStringList Session::getListeningIPs()
             || (listenIPv6 && (protocol == QAbstractSocket::IPv4Protocol)))
             continue;

+        // If an iface address pattern has been defined only allow
ips that match it to go
+        // through. Note that this will discard ipv6 addresses (the
wildcard matching is supported
+        // only for ipv4).
+        if (!ifaceAddrPattern.isEmpty() &&
!addressPatternMatches(ipString, ifaceAddrPattern))
+        {
+            logger->addMessage(tr("Blocked address %1 on pattern %2",
"Blocked address 124.5.1568.1 on pattern 10.*.*.*").arg(ipString,
ifaceAddrPattern), Log::INFO);
+            continue;
+        }
+        else
+        {
+            logger->addMessage(tr("Allowed address %1 on pattern %2",
"Allowed address 10.5.1568.1 on pattern 10.*.*.*").arg(ipString,
ifaceAddrPattern), Log::INFO);

Typo: 10.5.1568.1
------------------------------

In src/base/bittorrent/session.cpp
<https://github.com/qbittorrent/qBittorrent/pull/8159#discussion_r159062844>
:

> -        else
+        {
+            if (!ifaceAddrPattern.isEmpty() &&
!addressPatternMatches(ifaceAddr, ifaceAddrPattern))
+            {
+                logger->addMessage(tr("Configured network interface
address %1 doesn't match the given pattern.", "Configured network
interface address 124.5.1568.1 doesn't match the given
pattern.").arg(ifaceAddr), Log::CRITICAL);
+                IPs.append("127.0.0.1"); // Force listening to
localhost and avoid accidental connection that will expose user data.
+                return IPs;
+            }
+            else
+            {
+                IPs.append(ifaceAddr);
+            }
+        }
+        else if (!ifaceAddrPattern.isEmpty())
+        {
+            logger->addMessage(tr("No network interface/address is
specified, but address pattern is set.", "No network interface/address
is specified, but address pattern is set."), Log::CRITICAL);

The second argument of tr() is not needed because it's the same as the
first.

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub
<https://github.com/qbittorrent/qBittorrent/pull/8159#pullrequestreview-85955775>,
or mute the thread
<https://github.com/notifications/unsubscribe-auth/AA9weT_-B5OEUZO_A18a9Hfj85erBW8vks5tFO5ZgaJpZM4ROlQs>
.

Burgos commented:
Updated with the code style fixes.

Piccirello commented on src/base/bittorrent/session.cpp:
```diff
@@ -2412,13 +2413,32 @@ void Session::networkConfigurationChange(const QNetworkConfiguration& cfg)
+bool Session::addressPatternMatches(const QString& ipAddress, const QString& pattern)
+{
+    QStringList ipPieces = ipAddress.split(".");
+    QStringList patternPieces = pattern.split(".");
```
Both of these can be declared `const`.

Piccirello commented on src/base/bittorrent/session.cpp:
```diff
@@ -2412,13 +2413,32 @@ void Session::networkConfigurationChange(const QNetworkConfiguration& cfg)
+    QStringList ipPieces = ipAddress.split(".");
+    QStringList patternPieces = pattern.split(".");
+
+    if (ipPieces.size() != patternPieces.size()) {
```
You can remove the curly braces since this is a one-liner.

Piccirello commented on src/base/bittorrent/session.cpp:
```diff
@@ -2412,13 +2413,32 @@ void Session::networkConfigurationChange(const QNetworkConfiguration& cfg)
+        return false;
+    }
+
+    for (int i = 0; i < ipPieces.size(); i++) {
```
Same here- the curly braces can be removed from the `for` loop.

Piccirello commented on src/base/bittorrent/session.cpp:
```diff
@@ -2431,9 +2451,21 @@ const QStringList Session::getListeningIPs()
+            logger->addMessage(tr("No network interface/address is specified, but address pattern is set."), Log::CRITICAL);
+            IPs.append("127.0.0.1"); // Force listening to localhost and avoid accidental connection that will expose user data.
+        }
+        else // no constraints, listen on any interface/address
```
This `else` should actually have curly braces since the conditionals before it make use of them.

Burgos commented:
Updated with the const suggestion and more style fixes.

Piccirello commented on src/base/bittorrent/session.cpp:
```diff
@@ -2413,13 +2413,30 @@ void Session::networkConfigurationChange(const QNetworkConfiguration& cfg)
+        return false;
+
+    for (int i = 0; i < ipPieces.size(); i++)
+        if (patternPieces[i] != "*" && ipPieces[i] != patternPieces[i])
```
Please wrap each part of the conditional in parentheses

`if ((patternPieces[i] != "*") && (ipPieces[i] != patternPieces[i]))`

Piccirello commented on src/base/bittorrent/session.cpp:
```diff
@@ -2413,13 +2413,30 @@ void Session::networkConfigurationChange(const QNetworkConfiguration& cfg)
     }
 }
 
+bool Session::addressPatternMatches(const QString& ipAddress, const QString& pattern)
```
The ampersand should appear directly before the variable name, rather than after the variable type.

i.e. `const QString &ipAddress`

Piccirello commented on src/base/bittorrent/session.h:
```diff
@@ -581,6 +581,7 @@ namespace BitTorrent
         void applyBandwidthLimits();
         void processBannedIPs(libtorrent::ip_filter &filter);
         const QStringList getListeningIPs();
+        static bool addressPatternMatches(const QString& address, const QString& pattern);
```
Ampersand location here as well.

Piccirello commented:
Missed those last few on my initial review. Thanks for taking the time to work on this PR.

Burgos commented:
You're very welcome, thank you for the review :+1:. I've addressed your comments.

senposage commented:
this may close https://github.com/qbittorrent/qBittorrent/issues/8258

Burgos commented:
Yes, that looks like my use case as well. I would like to see this merged.

senposage commented:
can somebody link a build with this included so I can test my use case

Piccirello commented:
@senposage what OS + version do you use?

senposage commented:
anybody gonna remember to look at this ?
@Burgos @sledgehammer999

Burgos commented:
I'm sorry, I forgot! I have the binary, I'll attach it in an hour or two.

Burgos commented:
@senposage find the release build here. I can't attach it to this comment for some reason.

https://drive.google.com/file/d/17QQW8jRYkQa5FSz-SbDvoWrNbt05s8DK/view?usp=sharing

Burgos commented:
Thanks for the feedback! I would like to see this merged!

senposage commented:
or maby not it works on startup but if the vpn is disconnected/reconnected wo a client restart it still stalls
there needs to be a additional interface state detection logic if I goto into the options and change the bound ip via the drop down box and then change it back it starts right back up

glassez commented:
@Chocobo1, @thalieht, @Piccirello, what are your general thoughts about this feature?

NotTsunami commented:
Is this relevant still after #11592?

FranciscoPombal commented:
Abandoned and probably no longer relevant due to everything that has changed in the meantime.
